feat: add new model for LocalizedString type (#12)
commercetools/test-data: core/README.md

# `@commercetools-test-data/core`

This package provides the core functions to define the data models.

## Install

```bash
$ yarn add -D @commercetools-test-data/core
```

## Creating a new data model

> Models are treated packages. Each model will be published as a standalone NPM package.

Reviewer: What's the rationale behind defining the models as standalone NPM packages?

Author: Just so that you can pick what you want/need.

I suppose we can maybe group the models by domain or something instead.

- `commons`: for things like https://docs.commercetools.com/api/types
- `products`: for models related to the product domain.
- `accounts`: things like user, project, etc.

Would that be reasonable? cc @tdeekens

Reviewer 2: I like the idea but wonder it it's generally that useful however does not hurt. If we group by domain it might be helpful. You could ask the same for babel in parts I suppose.

To create a new model, add a new package in the `models` workspace with the following structure:

```
models/<package_name>
├── LICENSE
├── README.md
├── index.ts
├── package.json

> ├── src

Reviewer: What about `constants.ts` file that contains the constants?

Author: The structure here is mostly for the minimal setup. If certain models require more things, they can just do that.

Reviewer: Except that `constants` was an established convention, a unified way of defining constants in test-data. Maybe it's worth using the same convention here as well, to avoid future discussions.

Author: Does it mean that every model should have a `constants` file?

Reviewer: As far as I understood, yes, it behaves as a placeholder if there are no constants, just like the `presets` folder.

│   ├── builder.ts
│   ├── generator.ts
│   ├── index.ts
│   ├── presets
│   │   └── index.ts
│   ├── transformers.ts
│   └── types.ts
├── tsconfig.declarations.json
└── tsconfig.json
```

To define and build new models, you need to use the core package `@commercetools-test-data/core`.

### Builder

> A builder is the result of the data model. Builders allow composition with other models and to transform the data into different shapes, for example `rest` or `graphql`.

Reviewer 2: ```suggestion
A builder is the result of the data model. Builders allow compositions with other models and to transform the data into different shapes, for example `rest` or `graphql`.
```

When you want to get the generated data, you would **build** the model (builder).

Builders are essentially composed by a generator and (optionally) transformers.

### Generator

> A generator is what describes the initial model shape. Most of the fields can and should be initialized with some values. You can define random values using `fake` or static values.

Reviewer 2: ```suggestion
A generator is what describes the initial model shape and data. Most of the fields can and should be initialized with some values. You can define random values using `fake` or static values.
```

```ts
const generator = Generator<TAuthor>({

>   name: 'Author',

Reviewer: FYI: we can [not use](https://github.com/jackfranklin/test-data-bot#creating-your-first-builder) `name`.

> Feel free to use the name property if you like, but it's not used for anything in test-data-bot. It will probably get removed in a future major version.

Author: Ah thanks, I didn't know. I guess we can just drop it then.

@tdeekens was there a particular reason for us to define it?

Reviewer 2: None. Just carried forward from the first generator. We can drop it.

> ```ts
> const generator = Generator<TAuthor>({
>   name: 'Author',
>   fields: {
>     firstName: fake((f) => f.name.firstName()),
>     locale: 'en',
>   },
> });
> ```

Reviewer: ```suggestion
```ts
import { fake } from '@jackfranklin/test-data-bot';

const generator = Generator<TAuthor>({
  name: 'Author',
  fields: {
    firstName: fake((f) => f.name.firstName()),
    locale: 'en',
  },
});
```
```

Author: The `fake` is exported from our core package, so we should import from that. Thanks for the suggestion.

Reviewer: Yes I noticed that lately, the idea is to document the import.

> > The shape of the generator model is what usually is used in the UI and does not need transformation.

Reviewer: Not necessarily, we can use the REST or GraphQL transformation in the UI, I propose to rephrase it to

```suggestion
> The shape of the generator model is what is defined as a model in [commercetools HTTP API](https://docs.commercetools.com/api/).
```

Author: From what @tdeekens explained to me, the shape of the model should be based on what is used in the UI.

@tdeekens can you maybe clarify this?

Reviewer 2: Yes, I think it's reasonable that we say that the UI shape is the default. The other ones are API shapes. Basically we need to pick one as then we can map into each other. I think the default shape is the best candidate.

In this case, the generator can define the value of the nested field as `null`. Those nested models are then built in the builder and or the transformers.

For example, the author model can have a list of books:

```ts
const generator = Generator<TAuthor>({
  name: 'Author',
  fields: {
    firstName: fake((f) => f.name.firstName()),
    locale: 'en',
    books: null,
  },
});
```

Books are defined as another model:

```ts
const generator = Generator<TBook>({
  name: 'Book',
  fields: {
    title: fake((f) => f.random.words()),
  },
});
```

When defining the builder for the `Author`, we can then initialize the `books` field by referencing the `Book` model.

```ts
const Author = Builder<TAuthor>({
  generator,
  transformers,
}).books([Book.random()]);
```

At the same time, we also need to tell the transformers that the `books` field can be built. We do that by specifying the field in the `buildFields` option:

```ts
const transformers = {
  default: Transformer<TAuthor, TAuthor>('default', {
    buildFields: ['books'],
  }),
  graphql: Transformer<TAuthor, TAuthorGraphql>('graphql', {
    buildFields: ['books'],

>     addFields: () => ({
>       __typename: 'Author',
>     }),

Reviewer: Irrelevant.

```suggestion
```

Author: Why is it irrelevant? 🤔

Reviewer: The emphasis here is on `buildFields`, adding other stuff will make the reader destructed, in my opinion.

Author: I see. I'll put a `// ...` comment then to hint that other fields can be provided.

  }),
};
```

### Transformers

A transformer is used to transform the data defined in the generator (the initial model) to another shape.

There are 3 available transformer types:

- `default`: this is the "identity" transformer. You would use the `build()` method to get the generated data according to the initial shape.
- `graphql`: this is the transformer for GraphQL data shapes. You would use the `buildGraphql()` method to get the generated data transformed into a GraphQL shape.
- `rest`: this is the transformer for REST data shapes. You would use the `buildRest()` method to get the generated data transformed into a REST shape.

Defining a `graphql` and/or a `rest` transformer is optional and depends on what the specific data requirements are.

For example, for the `Author` we can define a `graphql` transformer like this:

```ts
const transformers = {

>   default: Transformer<TAuthor, TAuthor>('default', {
>     buildFields: ['books'],
>   }),

Reviewer: Irrelevant.

```suggestion
```

Author: Again, why is it irrelevant? 🤔

Reviewer: The emphasis understood from the following sentence is the GraphQL transformer. Adding other stuff can make the reader distracted, in my opinion.

> For example, for the `Author` we can define a `graphql` transformer like this:

Author: Yep ok 👌

    addFields: () => ({
      __typename: 'Author',
    }),
  }),
};
```

Then calling `Author.random().buildGraphql<TAuthorGraphql>()` will generate the data with the `__typename` field.

A transformer can be configured with different options:

- `buildFields`: a list of fields that are defined as other data models. This is necessary so that transformers of these nested models are also executed.
- `removeFields`: a list of fields that should be removed from the model.
- `addFields`: a function that should return an object with new fields that needs to be added to the model. The object is then merged with the initial data.
  The function gets an object argument with the initial `fields` object, in case it's needed to construct the new object.
- `replaceFields`: a function that should return the fully transformed value. This is useful for example when transforming data into different data types, for example an object into an array of objects.
  The function gets an object argument with the initial `fields` object, in case it's needed to construct the value.

> The `replaceFields` option is mutually exclusive in regards to the `addFields` and `removeFields` options.

### Presets

> Presets are pre-configured models (builders) that can be defined for specific use cases.

Reviewer: This may be outside the scope of the PR but it's worth mentioning. It's important to document how [presets should be exported as functions to ensure the randomness of data](https://github.com/commercetools/merchant-center-frontend/tree/master/packages-shared/test-data#presets-1).

Author: Ah good point, I'll try to document this as well.

### Presets

Presets are pre-configured models (builders) that can be defined for specific use cases and re-used.

For example:

```ts
// presets/author-with-one-book.ts
import type { TAuthorBuilder } from './types';
import * as Book from '../../book';
import Author from './builder';

const preset = (): TAuthorBuilder => Author.random().books([Book.random()]);
export default preset;

// presets/index.ts
export { default as authorWithOneBook } from './author-with-one-book';
```

Note that the preset exports a function. The function is useful in case the preset requires to use some random data, using the `faker` package. This ensures that on every function invocation, the returned data is indeed random.

For example:

```ts
// presets/author-with-one-book.ts
import type { TAuthorBuilder } from './types';
import * as faker from 'faker';
import * as Book from '../../book';
import Author from './builder';

const preset = (): TAuthorBuilder =>
  Author.random().books([Book.random().title(faker.random.words())]);
export default preset;

// presets/index.ts
export { default as authorWithOneBook } from './author-with-one-book';

> ```

Author: Here's the updated presets section

### Builder

A builder is the result of the data model. Builders allow composition with other models and to transform the data into different shapes, for example `rest` or `graphql`.
A builder is the result of the data model. Builders allow compositions with other models and to transform the data into different shapes, for example `rest` or `graphql`.

When you want to get the generated data, you would **build** the model (builder).

Builders are essentially composed by a generator and (optionally) transformers.

### Generator

A generator is what describes the initial model shape. Most of the fields can and should be initialized with some values. You can define random values using `fake` or static values.
A generator is what describes the initial model shape and data. Most of the fields can and should be initialized with some values. You can define random values using `fake` or static values.

```ts
import { Generator, fake } from '@commercetools-test-data/core';

const generator = Generator<TAuthor>({
  name: 'Author',
  fields: {
    firstName: fake((f) => f.name.firstName()),
    locale: 'en',
  },
});

> ```

Reviewer: I think this deletion was a typo.

A generator is what describes the initial model shape and data. Most of the fields can and should be initialized with some values. You can define random values using `fake` or static values.

> ````ts

Reviewer: Any reason for the additional backtick?

```suggestion
```ts
```
